<script>

    // if one err then , callback hell problem generated 
    //  callback function ---------------------------------------------
    first(2, function(firstValue, err) {
        if(!err) {
            second(firstValue, function(secondValue, err){
        if(!err){
            third(secondValue, function(thirdValue, err){
                if(!err){
                    console.log(thirdValue);
                }
            });
        }
            })
        }
    })

    function first(value , callback) {
        callback(value + 1, false);
    }

    function second (value , callback) {
        callback(value + 1, false);
    }

    function third(value , callback) {
        callback(value + 1, false);
    }

    // using Promise --------------------------------------------------------

    const promise = new Promise((resove, reject) => {
        resove(2);
    })

    promise.then(first).then(second).then(third).then(response => console.log(response));

    function first(value  ) {
        return value + 1;
    }
    function second (value  ) {
        return value + 1;
    }
    function third(value  ) {
        return value + 1;
    }


    // Promise in Details -------------------------------------------------------

    // let p = new Promise((resolve, reject) => {
    //     let a = 1 + 2;
    //     if(a == 2) {
    //         resolve('Success');
    //     } else {
    //         reject('Failed');
    //     }
    // });

    // p.then(response => console.log(response)).catch(err => console.log(err));

    // Promise All --------------------------------------------------------------

    // let p1 = new Promise((resolve, reject) => {
    //     setTimeout(() => {
    //         resolve('Success');
    //     }, 2000);
    // });

    // let p2 = new Promise((resolve, reject) => {
    //     setTimeout(() => {
    //         resolve('Success');
    //     }, 2000);
    // });

    // let p3 = new Promise((resolve, reject) => {
    //     setTimeout(() => {
    //         resolve('Success');
    //     }, 2000);
    // });

    // Promise.all([p1, p2, p3]).then(response => console.log(response));







    // syncronous code ---------------------------------------------------------
    
    // console.log(" I ")
    // console.log(" am ")
    // console.log(" syncronous ")
    // console.log(" code ")

    //    output : I 
    //             am
    //             syncronous
    //             code
    // everything executed in same order(one by one)
    // javascript run top to bottom


    // asyncronous code ---------------------------------------------------------
    console.log(" I ")
    setTimeout(() => {
        console.log(" am ")
    }, 2000);
    console.log(" asyncronous ")
    console.log(" code ")

    // output : I
    //          asyncronous
    //          code
    //          am


    // useMemo() and useCallback() -------------------------------------------------
    // useMemo()
    const sum = useMemo(() => {
        console.log("useMemo");
        return parseInt(inputValue1) + parseInt(inputValue2);
    }, [inputValue1, inputValue2]);
    
    // useCallback()
    const sampleFunc = useCallback(() => {
        console.log("inputValue1", inputValue1, "inputValue2", inputValue2);
    }, [inputValue1, inputValue2]);



// settTimout example ------------------------------------------------------------------
async function values () {
console.log("I")
console.log("am")
await new Promise((resolve, reject) => setTimeout(resolve,300))
.then(() => console.log("eating"));
console.log("this")
}

values()

// setTimeout example ----------------------------------------------
console.log("I")
console.log("am")
 new Promise((resolve, reject) => setTimeout(resolve(console.log("eating")),300))

console.log("this")


// console.log("I")
// console.log("am")
// setTimeout(async () => {
//   await console.log("eating")
// }, 3000)

// console.log("this")

</script>